// acorn-node/index.js
var acorn = require('acorn')
var xtend = require('xtend')

var CJSParser = acorn.Parser
  .extend(require('acorn-node/lib/bigint'))
  .extend(require('acorn-node/lib/class-fields'))
  .extend(require('acorn-node/lib/static-class-features'))
  .extend(require('acorn-node/lib/numeric-separator'))
  .extend(require('acorn-node/lib/dynamic-import').default)
var ESModulesParser = CJSParser
  .extend(require('acorn-node/lib/export-ns-from'))
  .extend(require('acorn-node/lib/import-meta'))

function mapOptions (opts) {
  if (!opts) opts = {}
  return xtend({
    ecmaVersion: 2020,
    allowHashBang: true,
    allowReturnOutsideFunction: true
  }, opts)
}

function getParser (opts) {
  if (!opts) opts = {}
  return opts.sourceType === 'module' ? ESModulesParser : CJSParser
}

module.exports = exports = xtend(acorn, {
  parse: function parse (src, opts) {
    return getParser(opts).parse(src, mapOptions(opts))
  },
  parseExpressionAt: function parseExpressionAt (src, offset, opts) {
    return getParser(opts).parseExpressionAt(src, offset, mapOptions(opts))
  },
  tokenizer: function tokenizer (src, opts) {
    return getParser(opts).tokenizer(src, mapOptions(opts))
  }
})
